pkg/kubelet/nodeconfig/controller.go

```diff
@@ -0,0 +1,270 @@
+		if cc.initConfig != nil {
+			cc.setConfigOK("using current (init)",
+				"current is set to the local default, and an init config was provided", apiv1.ConditionTrue)
+			return cc.initConfig
+		}
+		cc.setConfigOK("using current (default)",
+			"current is set to the local default, and no init config was provided", apiv1.ConditionTrue)
+		return cc.defaultConfig
+	} // Assert: we will not use the init or default configurations, unless we roll back to lkg; curUID is a real UID
+
+	// check whether the current config is marked bad
+	if bad, entry := cc.isBadConfig(curUID); bad {
+		infof("current config %q was marked bad for reason %q at time %q", curUID, entry.reason, entry.time)
+		return cc.lkgRollback(entry.reason, apiv1.ConditionFalse)
+	}
+
+	// TODO(mtaufen): consider re-verifying integrity and re-attempting download when a load/verify/parse/validate
+	// error happens outside trial period, we already made it past the trial so it's probably filesystem corruption
+	// or something else scary
+
+	// load the current config
+	toVerify, err := cc.loadCheckpoint(curSymlink)
+	if err != nil {
+		// TODO(mtaufen): rollback and mark bad for now, but this could reasonably be handled by re-attempting a download,
+		// it probably indicates some sort of corruption
+		return cc.badRollback(curUID, fmt.Sprintf("failed to load current (UID: %q)", curUID), fmt.Sprintf("error: %v", err))
+	}
+
+	// verify the integrity of the configuration we just loaded
+	toParse, err := toVerify.verify()
+	if err != nil {
+		return cc.badRollback(curUID, fmt.Sprintf("failed to verify current (UID: %q)", curUID), fmt.Sprintf("error: %v", err))
+	}
+
+	// parse the configuration we just loaded into a KubeletConfiguration
+	cur, err := toParse.parse()
+	if err != nil {
+		return cc.badRollback(curUID, fmt.Sprintf("failed to parse current (UID: %q)", curUID), fmt.Sprintf("error: %v", err))
+	}
+
+	// validate current config
+	if err := validateConfig(cur); err != nil {
+		return cc.badRollback(curUID, fmt.Sprintf("failed to validate current (UID: %q)", curUID), fmt.Sprintf("error: %v", err))
+	}
+
+	// check for crash loops if we're still in the trial period
+	if cc.curInTrial(cur.ConfigTrialDuration.Duration) {
+		if cc.curCrashLooping(*cur.CrashLoopThreshold) {
+			return cc.badRollback(curUID, fmt.Sprintf("current failed trial period due to crash loop (UID %q)", curUID), "")
+		}
+	} else if !cc.curIsLkg() {
+		// when the trial period is over, the current config becomes the last-known-good
+		cc.setCurAsLkg()
+	}
+
+	// update the status to note that we will use the current config
+	cc.setConfigOK(fmt.Sprintf("using current (UID: %q)", curUID), "passed all checks", apiv1.ConditionTrue)
+
+	return cur
+}
+
+// initSync launches the sync loop that watches the API server for new configuration
+func (cc *NodeConfigController) initSync() {
+	if cc.client != nil {
+		infof("starting sync loop")
+		go func() {
+			fieldselector := fmt.Sprintf("metadata.name=%s", cc.nodename)
```

Comment:
HandleError takes an error as an argument though. Should I call it with a nil error? I'm not sure what the point of this would be, since it just returns immediately when passed a nil error (I'm looking in `kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go`).